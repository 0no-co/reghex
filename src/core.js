const isStickySupported = typeof /./g.sticky === 'boolean';

export const _pattern = (input) => {
  if (typeof input === 'function') return input;

  const source = typeof input !== 'string' ? input.source : input;
  return isStickySupported
    ? new RegExp(source, 'y')
    : new RegExp(`^(?:${source})`, 'g');
};

export const _substr = (state, pattern) => {
  const end = state.index + pattern.length;
  const sub = state.input.slice(state.index, end);
  if (sub === pattern) {
    state.index = end;
    return sub;
  }
};

export const _exec = (state, pattern) => {
  if (typeof pattern === 'function') return pattern();

  let match;
  if (isStickySupported) {
    pattern.lastIndex = state.index;
    match = pattern.exec(state.input);
    state.index = pattern.lastIndex;
  } else {
    pattern.lastIndex = 0;
    match = pattern.exec(state.input.slice(state.input));
    state.index += pattern.lastIndex;
  }

  return match && match[0];
};

export const tag = (array, tag) => {
  array.tag = tag;
  return array;
};

export const parse = (pattern) => (input) => {
  const state = { input, index: 0 };
  return pattern(state);
};

export const match = (_name) => {
  throw new TypeError(
    'This match() function was not transformed. ' +
      'Ensure that the Babel plugin is set up correctly and try again.'
  );
};
