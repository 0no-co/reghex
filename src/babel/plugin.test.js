import { transform } from '@babel/core';
import reghexPlugin from './plugin';

it('works with standard features', () => {
  const code = `
    import match from 'reghex/macro';

    const node = match('node')\`
      \${1}+ | \${2}+ (\${3} ( \${4}? \${5} ) )*
    \`;
  `;

  expect(
    transform(code, { babelrc: false, presets: [], plugins: [reghexPlugin] })
      .code
  ).toMatchSnapshot();
});

it('works with local recursion', () => {
  // NOTE: A different default name is allowed
  const code = `
    import match_rec, { tag } from 'reghex';

    const inner = match_rec('inner')\`
      \${/inner/}
    \`;

    const node = match_rec('node')\`
      \${inner}
    \`;
  `;

  expect(
    transform(code, { babelrc: false, presets: [], plugins: [reghexPlugin] })
      .code
  ).toMatchSnapshot();
});

it('works with transform functions', () => {
  const code = `
    import match from 'reghex';

    const first = match('inner', x => x)\`\`;

    const transform = x => x;
    const second = match('node', transform)\`\`;
  `;

  expect(
    transform(code, { babelrc: false, presets: [], plugins: [reghexPlugin] })
      .code
  ).toMatchSnapshot();
});

it('works with non-capturing groups', () => {
  const code = `
    import match from 'reghex';

    const node = match('node')\`
      \${1} (\${2} | (?: \${3})+)
    \`;
  `;

  expect(
    transform(code, { babelrc: false, presets: [], plugins: [reghexPlugin] })
      .code
  ).toMatchSnapshot();
});

it('works together with @babel/plugin-transform-modules-commonjs', () => {
  const code = `
    import match from 'reghex';

    const node = match('node')\`
      \${1} \${2}
    \`;
  `;

  expect(
    transform(code, {
      babelrc: false,
      presets: [],
      plugins: [
        reghexPlugin,
        [
          '@babel/plugin-transform-modules-commonjs',
          {
            noInterop: true,
            loose: true,
          },
        ],
      ],
    }).code
  ).toMatchSnapshot();
});
