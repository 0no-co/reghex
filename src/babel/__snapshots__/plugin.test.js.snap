// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works together with @babel/plugin-transform-modules-commonjs 1`] = `
"\\"use strict\\";

var _reghex = require(\\"reghex\\");

var _node_expression = (0, _reghex._pattern)(1),
    _node_expression2 = (0, _reghex._pattern)(2);

const node = function _node(state) {
  var last_index = state.index;
  var match,
      node = [];

  if (match = (0, _reghex._exec)(state, _node_expression)) {
    node.push(match);
  } else {
    state.index = last_index;
    return;
  }

  if (match = (0, _reghex._exec)(state, _node_expression2)) {
    node.push(match);
  } else {
    state.index = last_index;
    return;
  }

  return (0, _reghex.tag)(node, 'node');
};"
`;

exports[`works with local recursion 1`] = `
"import { tag, _exec, _substr, _pattern } from 'reghex';

const inner = function _inner(state) {
  var last_index = state.index;
  var match,
      node = [];

  if (match = _substr(state, \\"inner\\")) {
    node.push(match);
  } else {
    state.index = last_index;
    return;
  }

  return tag(node, 'inner');
};

const node = function _node(state) {
  var last_index = state.index;
  var match,
      node = [];

  if (match = inner(state)) {
    node.push(match);
  } else {
    state.index = last_index;
    return;
  }

  return tag(node, 'node');
};"
`;

exports[`works with non-capturing groups 1`] = `
"import { _exec, _substr, _pattern, tag as _tag } from 'reghex';

var _node_expression = _pattern(1),
    _node_expression2 = _pattern(2),
    _node_expression3 = _pattern(3);

const node = function _node(state) {
  var last_index = state.index;
  var match,
      node = [];

  if (match = _exec(state, _node_expression)) {
    node.push(match);
  } else {
    state.index = last_index;
    return;
  }

  var length_0 = node.length;

  alternation_1: {
    block_1: {
      var index_1 = state.index;

      if (match = _exec(state, _node_expression2)) {
        node.push(match);
      } else {
        node.length = length_0;
        state.index = index_1;
        break block_1;
      }

      break alternation_1;
    }

    loop_1: for (var iter_1 = 0; true; iter_1++) {
      var index_1 = state.index;

      if (!_exec(state, _node_expression3)) {
        if (iter_1) {
          state.index = index_1;
          break loop_1;
        }

        node.length = length_0;
        state.index = last_index;
        return;
      }
    }
  }

  return _tag(node, 'node');
};"
`;

exports[`works with standard features 1`] = `
"import { _exec, _substr, _pattern, tag as _tag } from \\"reghex\\";

var _node_expression = _pattern(1),
    _node_expression2 = _pattern(2),
    _node_expression3 = _pattern(3),
    _node_expression4 = _pattern(4),
    _node_expression5 = _pattern(5);

const node = function _node(state) {
  var last_index = state.index;
  var match,
      node = [];

  block_0: {
    var index_0 = state.index;

    loop_0: for (var iter_0 = 0; true; iter_0++) {
      var index_0 = state.index;

      if (match = _exec(state, _node_expression)) {
        node.push(match);
      } else {
        if (iter_0) {
          state.index = index_0;
          break loop_0;
        }

        state.index = index_0;
        break block_0;
      }
    }

    return _tag(node, 'node');
  }

  loop_0: for (var iter_0 = 0; true; iter_0++) {
    var index_0 = state.index;

    if (match = _exec(state, _node_expression2)) {
      node.push(match);
    } else {
      if (iter_0) {
        state.index = index_0;
        break loop_0;
      }

      state.index = last_index;
      return;
    }
  }

  loop_0: while (true) {
    var index_0 = state.index;
    var length_0 = node.length;

    if (match = _exec(state, _node_expression3)) {
      node.push(match);
    } else {
      node.length = length_0;
      state.index = index_0;
      break loop_0;
    }

    var index_2 = state.index;

    if (match = _exec(state, _node_expression4)) {
      node.push(match);
    } else {
      state.index = index_2;
    }

    if (match = _exec(state, _node_expression5)) {
      node.push(match);
    } else {
      node.length = length_0;
      state.index = index_0;
      break loop_0;
    }
  }

  return _tag(node, 'node');
};"
`;

exports[`works with transform functions 1`] = `
"import { _exec, _substr, _pattern, tag as _tag } from 'reghex';

var _inner_transform = x => x;

const first = function _inner(state) {
  var last_index = state.index;
  var match,
      node = [];
  return _inner_transform(_tag(node, 'inner'));
};

const transform = x => x;

const second = function _node(state) {
  var last_index = state.index;
  var match,
      node = [];
  return transform(_tag(node, 'node'));
};"
`;
